src/openvic-simulation/types/Colour.hpp

```diff
@@ -1,42 +1,218 @@
 #pragma once
 
 #include <algorithm>
+#include <array>
+#include <charconv>
+#include <cmath>
+#include <compare>
 #include <cstdint>
 #include <iomanip>
+#include <limits>
+#include <span>
 #include <sstream>
 #include <string>
+#include <string_view>
+#include <type_traits>
 
 namespace OpenVic {
+	template<typename ValueT, typename ComponentReturnT>
+	struct colour_traits {
+		using value_type = ValueT;
+		using component_return_type = ComponentReturnT;
+		ValueT null = 0;
+		ValueT component = 0xFF;
+		ValueT blue_shift = 0;
+		ValueT green_shift = 8;
+		ValueT red_shift = 16;
+		ValueT alpha_shift = 24;
+		ValueT max_rgb = (component << red_shift) | (component << green_shift) | (component << blue_shift);
+		ValueT max_argb = (component << alpha_shift) | max_rgb;
+	};
+
 	/* Colour represented by an unsigned integer, either 24-bit RGB or 32-bit ARGB. */
-	using colour_t = uint32_t;
-
-	/* When colour_t is used as an identifier, NULL_COLOUR is disallowed
-	 * and should be reserved as an error value.
-	 * When colour_t is used in a purely graphical context, NULL_COLOUR
-	 * should be allowed.
-	 */
-	static constexpr colour_t NULL_COLOUR = 0, COLOUR_COMPONENT = 0xFF;
-	static constexpr colour_t MAX_COLOUR_RGB = 0xFFFFFF, MAX_COLOUR_ARGB = 0xFFFFFFFF;
-
-	constexpr colour_t float_to_colour_byte(float f, float min = 0.0f, float max = 1.0f) {
-		return static_cast<colour_t>(std::clamp(min + f * (max - min), min, max) * 255.0f);
-	}
+	template<typename ValueT, typename ComponentReturnT, colour_traits Traits = colour_traits<ValueT, ComponentReturnT> {}>
+	struct basic_colour_t {
+		static constexpr auto traits = Traits;
+		using value_type = typename decltype(traits)::value_type;
+		using component_return_type = typename decltype(traits)::component_return_type;
 
-	constexpr colour_t fraction_to_colour_byte(int n, int d, float min = 0.0f, float max = 1.0f) {
-		return float_to_colour_byte(static_cast<float>(n) / static_cast<float>(d), min, max);
-	}
+		value_type value;
 
-	constexpr colour_t float_to_alpha_value(float a) {
-		return float_to_colour_byte(a) << 24;
-	}
+		/* When colour_t is used as an identifier, NULL_COLOUR is disallowed
+		 * and should be reserved as an error value.
+		 * When colour_t is used in a purely graphical context, NULL_COLOUR
+		 * should be allowed.
+		 */
+		static constexpr value_type null = traits.null;
+		static constexpr value_type component = traits.component;
+		static constexpr value_type max_rgb = traits.max_rgb;
+		static constexpr value_type max_argb = traits.max_argb;
 
-	constexpr float colour_byte_to_float(colour_t colour) {
-		return std::clamp(static_cast<float>(colour) / 255.0f, 0.0f, 1.0f);
-	}
```

Comment:
We need a variant of this for converting `colour_t` to `godot::Color` in the extension - maybe a static component type -> float conversion function, then non-static `redf`, `greenf`, `bluef`, `alphaf` getters using it?